Fix #9 and #43 (port oscoin multibase abstraction to this project)

adinapoli-mndc opened the pull request:
Currently a work in progress, but the key idea is that I have used a `Symbol` instead of a `Nat` to represent encodings which have the same "number" (i.e. 32 vs 32z, 58btc vs 58flickr etc).

**NOTE:** The "NG" module was added only to simplify the review process. Once assessed this is going into the right direction, I will ultimately replace the `Multibase` one with this.

adinapoli-mndc commented:
```

exe:git-remote-ipfs from git-remote-ipfs-0.1.0.0).
cabal: Failed to build git-remote-ipfs-0.1.0.0 (which is required by
see the error.
Warning: Some package(s) failed to build. Try rerunning with -j1 if you can't
| ^^^^^^^^^^^^^^^^
56 | import Servant.Client (ServantError(..))
|
Module ‘Servant.Client’ does not export ‘ServantError(..)’
src/Network/IPFS/Git/RemoteHelper/Client.hs:56:34: error:
```

🤔

adinapoli-mndc commented:
@kim , this is now ready for review (I am off on Monday, so you might have to wait till Tuesday before cursing me 😛 ). At the end of the day the design revealed to be slightly more boilerplat-y than I would have hoped, mostly because my initial type signature `decode :: Multibase -> Either String (AtBase b)` was a bit of a lie, as in practice that would mean simply removing the `code` from the input bytestring but it's not that useful -- in practice, you also want to decode, hehe (as per original behaviour).

However, this meant that now I had no mention of the type variable `b` in the type signature, which meant I could invoke `fromCode` (as part of typeclass) anymore, which meant in turn `fromCode` _had_ to live as a standalone (verbose) function.

An alternative would be to tag `Multibase` with a phantom type `b` as in `Multibase b`, but this is quite heavy I think, and would later on hamper our ability to treat is as an opaque blob (and implement `uknownMultibase`).

kim commented on multibase/src/Data/ByteString/BaseN.hs:
```diff
@@ -0,0 +1,932 @@
+{-# INLINE decodeBase32hex #-}
+
+-- |
+-- fromAtBase . encodeBase32hexupper $ "hello world"
```
```suggestion
-- >>> fromAtBase . encodeBase32hexupper $ "hello world"
```

kim commented on multibase/src/Data/ByteString/BaseN.hs:
```diff
@@ -0,0 +1,932 @@
+-- Base 32 ---------------------------------------------------------------------
+
+-- |
+-- fromAtBase . encodeBase32hex $ "hello world"
```
Did I write these doctests?

```suggestion
-- >>> fromAtBase . encodeBase32hex $ "hello world"
```

adinapoli-mndc replied:
I _think_ you did, but I did add the `fromAtBase`. Some of them were imported from `Multibase.hs`, others from `BaseN.hs`, so paternity is unclear 😂

kim commented on multihash-serialise/src/Codec/Serialise/Multihash.hs:
```diff
@@ -27,12 +29,30 @@ import qualified Codec.Serialise as CBOR
 import           Data.Proxy (Proxy(..))
 
+-- | Decode a 'C.Digest' from a multihash- and base-n-encoded 'ByteString'.
+decodeAtBase
```
I guess this doesn't _really_ belong here: either we're in the mülti wörld, or bareback bytestrings.

adinapoli-mndc replied:
Yeah, I feel you. However, this was the last little shim I had to write in `oscoin` to be able to tie the knot between these library and application code. Mind, I think the answer "this shim belongs into application code" is a perfectly reasonable answer, if we don't think this has enough general-value to be outsourced into library code 😉

kim commented on multibase/src/Data/ByteString/Multibase.hs:
```diff
@@ -1,13 +1,11 @@
-{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE TypeApplications           #-}
 
 -- | Kitchen sink base-N encoding and decoding of strict 'ByteStrings'.
```
```suggestion
-- |
-- Copyright   : 2019 Monadic GmbH
-- License     : BSD3
-- Maintainer  : kim@monadic.xyz, alfredo@monadic.xyz, team@monadic.xyz
-- Stability   : provisional
-- Portability : non-portable (GHC extensions)
--
-- <https://github.com/multiformats/multibase multibase> encoding of strict 'ByteString's.
--
```

kim commented on multibase/src/Data/ByteString/BaseN.hs:
```diff
@@ -0,0 +1,932 @@
+{-# LANGUAGE TupleSections              #-}
+{-# LANGUAGE TypeApplications           #-}
+
+-- | Kitchen sink base-N encoding and decoding of strict 'ByteStrings'.
```
```suggestion
-- |
-- Copyright   : 2019 Monadic GmbH
-- License     : BSD3
-- Maintainer  : kim@monadic.xyz, alfredo@monadic.xyz, team@monadic.xyz
-- Stability   : provisional
-- Portability : non-portable (GHC extensions)
--
-- Type-safe kitchen sink base-N encoding and decoding of strict 'ByteString's.
--
```

adinapoli-mndc replied:
Good man. Always meant to ask -- do you use some code snippet in your editor to write up those headers?

kim commented on multibase/src/Data/ByteString/BaseN.hs:
```diff
@@ -0,0 +1,932 @@
+{-# LANGUAGE ConstraintKinds            #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
```
British spelling was introduced in [GHC 8.6.1](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/8.6.1-notes.html#compiler), see https://travis-ci.org/oscoin/ipfs/jobs/516248105#L351

🤣

adinapoli-mndc replied:
Lol, did I unconsciously plugged a brit-spelled lang extension into the codebase? @cloudhead 's heart attack in 3..2..1 😂

adinapoli-mndc commented:
@kim I also now added support for the `base32z` encoding, which is where the `Symbol` approach starts to pay off.

Unfortunately I couldn't plug this back into `oscoin` just yet as `radicle` would need to be upgraded to support the change of interface (now `decodeBase64` doesn't live in `Multibase` anymore, but in `BaseN`). Doing that might be wasted effort considering our plans to remove `radicle` from `oscoin` anyway.

The issue with `decodeAtBase` and `encodeAtBase` is still open, but I am happy to scrap those 2 functions from `multihash-serialise`.

kim commented:
> The issue with decodeAtBase and encodeAtBase is still open

Do we have reasons to *not* do multibase(multihash(x))? I forget

adinapoli-mndc commented:
> Do we have reasons to *not* do multibase(multihash(x))? I forget

To be honest with you I don't know -- these shims were ported from `oscoin` and were already there by the time I refactored the initial strand of code 😅

adinapoli-mndc commented:
@kim _monsieur_ (I am not really here today 😁 ) , I have fixed CI and removed the `encode/decodeAtBase` helpers, as I don't mind keeping them in application code. Do you think you can merge this guy now or is there anything else you would like to see?

kim commented:
@adinapoli-mndc A rebase would be nice :)

adinapoli commented:
@kim Done! 😉

adinapoli-mndc commented:
Damn, I did it again, I commented with my doppelganger! 💔
